agent/config/builder.go

```diff
@@ -1918,6 +1927,136 @@ func (b *Builder) isUnixAddr(a net.Addr) bool {
 	return a != nil && ok
 }
 
+func (b *Builder) autoConfigVal(raw AutoConfigRaw) AutoConfig {
+	var val AutoConfig
+
+	val.Enabled = b.boolValWithDefault(raw.Enabled, false)
+	val.IntroToken = b.stringVal(raw.IntroToken)
+	val.IntroTokenFile = b.stringVal(raw.IntroTokenFile)
+	// These can be go-discover values and so don't have to resolve fully yet
+	val.ServerAddresses = b.expandAllOptionalAddrs("auto_config.server_addresses", raw.ServerAddresses)
+	val.DNSSANs = raw.DNSSANs
+
+	for _, i := range raw.IPSANs {
+		ip := net.ParseIP(i)
+		if ip == nil {
+			b.warn(fmt.Sprintf("Cannot parse ip %q from auto_config.ip_sans", i))
+			continue
+		}
+		val.IPSANs = append(val.IPSANs, ip)
+	}
+
+	val.Authorizer = b.autoConfigAuthorizerVal(raw.Authorizer)
+
+	return val
+}
+
+func (b *Builder) autoConfigAuthorizerVal(raw AutoConfigAuthorizerRaw) AutoConfigAuthorizer {
+	var val AutoConfigAuthorizer
+
+	val.Enabled = b.boolValWithDefault(raw.Enabled, false)
+	val.ClaimAssertions = raw.ClaimAssertions
+	val.AllowReuse = b.boolValWithDefault(raw.AllowReuse, false)
+	val.AuthMethod = structs.ACLAuthMethod{
+		Name: "Auto Config Authorizer",
+		Type: "jwt",
+		// TODO (autoconf) - Configurable token TTL
+		MaxTokenTTL:    72 * time.Hour,
+		EnterpriseMeta: *structs.DefaultEnterpriseMeta(),
+		Config: map[string]interface{}{
+			"JWTSupportedAlgs":     raw.JWTSupportedAlgs,
+			"BoundAudiences":       raw.BoundAudiences,
+			"ClaimMappings":        raw.ClaimMappings,
+			"ListClaimMappings":    raw.ListClaimMappings,
+			"OIDCDiscoveryURL":     b.stringVal(raw.OIDCDiscoveryURL),
+			"OIDCDiscoveryCACert":  b.stringVal(raw.OIDCDiscoveryCACert),
+			"JWKSURL":              b.stringVal(raw.JWKSURL),
+			"JWKSCACert":           b.stringVal(raw.JWKSCACert),
+			"JWTValidationPubKeys": raw.JWTValidationPubKeys,
+			"BoundIssuer":          b.stringVal(raw.BoundIssuer),
+			"ExpirationLeeway":     b.durationVal("auto_config.authorizer.expiration_leeway", raw.ExpirationLeeway),
+			"NotBeforeLeeway":      b.durationVal("auto_config.authorizer.not_before_leeway", raw.NotBeforeLeeway),
+			"ClockSkewLeeway":      b.durationVal("auto_config.authorizer.clock_skew_leeway", raw.ClockSkewLeeway),
+		},
+		// should be unnecessary as we aren't using the typical login process to create tokens but this is our
+		// desired mode regardless so if it ever did matter its probably better to be explicit.
+		TokenLocality: "local",
+	}
+
+	return val
+}
+
+func (b *Builder) validateAutoConfig(rt RuntimeConfig) error {
+	autoconf := rt.AutoConfig
+
+	if autoconf.Enabled {
```

Comment:
This could be a guard `if !autoconf.Enabled { return b.validateAutoConfigAuthorizer(rt) }`.

Would it be appropriate to warn in the case where autoconf is disabled but the user has specified some config in this section?

Reply:
I don't think its a problem if they haven't enabled but provided some configuration. I am thinking specifically of the case where they had enabled it but now want to disable. In that case they might just flip from enabled = true to enabled = false. Requiring an extra step to remove/comment the rest of the configuration seems like an unnecessary annoyance.

Reply:
As for the guard I refactored things a little so there is less nesting.